### PR TITLE
Upgrade Logging facility

### PR DESCRIPTION
**Pull request by chick**
Make thread-safe
Make logging by package name work
Use caching of class names to level for performance
Make some tests to show this working

### CODE REVIEW

File: src/main/scala/logger/Logger.scala [hunk 1]
```diff
@@ -36,45 +38,103 @@ trait LazyLogging {
...
+        }
+      }
+      state.classToLevelCache(className) = matchPathToFindLevel(packageNameList)
+      state.classToLevelCache(className)
+    })
+    levelForThisClassName >= level
+  }
+
+  //scalastyle:off regex
   def showMessage(level: LogLevel.Value, className: String, message: => String): Unit = {
```
Comment (azidar) [inline src/main/scala/logger/Logger.scala]: Comment

**Comment by chick** [comment]
@azidar Took longer than anticipated, I hope this works for you now

File: src/main/scala/logger/Logger.scala [hunk 2]
```diff
@@ -36,45 +38,103 @@ trait LazyLogging {
   val logger = new Logger(this.getClass.getName)
 }
 
+class LoggerState {
```
Comment (azidar) [inline src/main/scala/logger/Logger.scala]: and make private please

**Comment by azidar** [comment]
Looks like the default logger mode isn't set right - the travis build fails because it ran out of space to hold the log, as all tests are printing a lot.

**Comment by chick** [comment]
@azidar I have pushed another fix, we should discuss in morning

**Comment by chick** [comment]
@azidar I think the current commit of this PR addresses the issues we went over.  The chisel3 tutorial branch uses the latest change to show a straightforward way of capturing logging.

**Comment by shunshou** [comment]
Does  commonOptions = CommonOptions(globalLogLevel = LogLevel.Debug)
work?

It seems like I'm not getting any useful logger info -- even tho I was able to in the past?

**Comment by shunshou** [comment]
@chick -- seems like chisel3.driver passes logger info down correctly, but dsptools.driver (and I assume iotesters.driver) loses the logger option...

**Comment by chick** [comment]
@shunshou I don't see where these would be lost when I take a quick look at 
dspTools.Driver -> iotestesters.Driver -> chisel3.Driver.  But I might be missing something
BTW, where does the code ```commonOptions = CommonOptions(globalLogLevel = LogLevel.Debug)``` come from.

**Comment by chick** [comment]
@shunshou I think I understand the problem.  The Driver.execute methods of dsptools and iotesters and chisel3 will need a change to have them automatically set the logger options.  It's simple, it just needs to wrap the execute method's code in a Logger.makeScope.  If you think this might be the problem I can create PR's to introduce the compatibility change

**Pull request merged**